OpenRA.Mods.Cnc/Traits/MadTank.cs

```diff
@@ -162,68 +139,121 @@ string IOrderVoice.VoicePhraseForOrder(Actor self, Order order)
-				{
-					// Use .FromPos since this actor is killed. Cannot use Target.FromActor
-					info.DetonationWeaponInfo.Impact(Target.FromPos(self.CenterPosition), self, Enumerable.Empty<int>());
-				}
-
-				self.Kill(self, info.DamageTypes);
-			});
+				self.SetTargetLine(order.Target, Color.Red);
+				self.QueueActivity(order.Queued, new DetonationSequence(self, this, order.Target));
+			}
+			else if (order.OrderString == "Detonate")
+				self.QueueActivity(order.Queued, new DetonationSequence(self, this));
 		}
 
-		void EjectDriver()
+		class DetonationSequence : Activity
 		{
-			var driver = self.World.CreateActor(info.DriverActor.ToLowerInvariant(), new TypeDictionary
+			readonly Actor self;
+			readonly MadTank mad;
+			readonly WithFacingSpriteBody wfsb;
+			readonly ScreenShaker screenShaker;
+			readonly bool assignTargetOnFirstRun;
+
+			int ticks;
+			bool initiated;
+			Target target;
+
+			public DetonationSequence(Actor self, MadTank mad)
+				: this(self, mad, Target.Invalid)
 			{
-				new LocationInit(self.Location),
-				new OwnerInit(self.Owner)
-			});
-			var driverMobile = driver.TraitOrDefault<Mobile>();
-			if (driverMobile != null)
-				driverMobile.Nudge(driver, driver, true);
-		}
+				assignTargetOnFirstRun = true;
+			}
 
-		void StartDetonationSequence()
-		{
-			if (deployed)
-				return;
-
-			if (conditionManager != null && !string.IsNullOrEmpty(info.DeployedCondition))
-				conditionManager.GrantCondition(self, info.DeployedCondition);
-
-			self.World.AddFrameEndTask(w => EjectDriver());
-			if (info.ThumpSequence != null)
-				wfsb.PlayCustomAnimationRepeating(self, info.ThumpSequence);
-			deployed = true;
-			self.QueueActivity(new Wait(info.ChargeDelay, false));
-			self.QueueActivity(new CallFunc(() => Game.Sound.Play(SoundType.World, info.ChargeSound, self.CenterPosition)));
-			self.QueueActivity(new Wait(info.DetonationDelay, false));
-			self.QueueActivity(new CallFunc(() => Game.Sound.Play(SoundType.World, info.DetonationSound, self.CenterPosition)));
-			self.QueueActivity(new CallFunc(Detonate));
-		}
+			public DetonationSequence(Actor self, MadTank mad, Target target)
+			{
+				this.self = self;
+				this.mad = mad;
+				this.target = target;
+				wfsb = self.Trait<WithFacingSpriteBody>();
+				screenShaker = self.World.WorldActor.Trait<ScreenShaker>();
+			}
 
-		void IResolveOrder.ResolveOrder(Actor self, Order order)
-		{
-			if (order.OrderString == "DetonateAttack")
+			protected override void OnFirstRun(Actor self)
 			{
-				if (!order.Queued)
-					self.CancelActivity();
+				if (assignTargetOnFirstRun)
+					target = Target.FromCell(self.World, self.Location);
+			}
 
-				self.SetTargetLine(order.Target, Color.Red);
-				self.QueueActivity(new MoveAdjacentTo(self, order.Target, targetLineColor: Color.Red));
-				self.QueueActivity(new CallFunc(StartDetonationSequence));
+			public override bool Tick(Actor self)
+			{
+				if ((self.Location - target.Actor.Location).LengthSquared > 3)
```

Comment:
If we want to be more relaxed about exactly where we deploy we could instead define something like `public readonly WDist TargetDetonationRange = WDist.FromCells(3);` in the `MADTankInfo` and then use `if (target.Type != TargetType.Invalid && target.IsInRange(self.CenterPosition, mad.info.TargetDetonationRange))`

This is probably a better solution, but I haven't tested it.

Reply:
The second suggestion sounds better to me.

Reply:
Thinking about this further, is there any reason why we couldn't/shouldn't use the range defined on `DetonationWeapon`? This would be better than defining a separate `TargetDetonationRange`.

Reply:
I would presume that when you use the madtank you want to take out as many buildings at once, so you would choose a target in the middle of a clump of buildings. Deploying as soon as the target is on the edge of the detonation radius doesn't seem very effective.

Especially if you want to later on add support for force-attack commands, it seems most intuitive to me if the madtank would actually move to the ordered tile instead of somewhere on the edge of the range-circle. So I think that the first suggestion makes more sense from a practical viewpoint.